### PR TITLE
Add precision@K to GAME, and address some related issues.

### PR DESCRIPTION
**Pull request by XianXing**
A few things to watch out for in this PR:

0. Precision@K is now ready to be used as an evaluator in GAME, similar to other metrics such as AUC and RMSE.

1. ```GameDatum``` now hold all types of Ids other than just random effect ids. For example, ```uid``` used for some scoring downstream applications and ```documentId``` or ```queryId``` used to compute precision@K metric (or replay metric in the future). This change simplifies the data processing logic for GAME's scoring and evaluation modules.

2. GAME now supports running multiple evaluators at the same time, for example, the users can specify precision@1 + precision@3 + precision@10 + AUC as evaluators at the same time, and GAME will spit out four evaluation metrics per iteration - given the validation data is specified by the user. Motivation for this feature is stated in #164.

3. Minor issues like #165 is addressed, along with some minor Scala code style fixes - sorry, I just can't help fixing them when seeing those warning in my Intellij.

### CODE REVIEW

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/EvaluatorType.scala [hunk 1]
```diff
@@ -14,7 +14,67 @@
...
+    case AUC.name => AUC
+    case RMSE.name => RMSE
+    case LogisticLoss.name | "LOGISTICLOSS" => LogisticLoss
+    case PoissonLoss.name | "POISSONLOSS" => PoissonLoss
+    case SmoothedHingeLoss.name | "SMOOTHEDHINGELOSS" => SmoothedHingeLoss
+    case SquaredLoss.name | "SQUAREDLOSS" => SquaredLoss
+    case PrecisionAtK.precisionAtKPattern(k, _) =>
+      val PrecisionAtK.precisionAtKPattern(_, documentId) = name.trim
+      PrecisionAtK(k.toInt, documentId)
+    case _ => throw new UnsupportedOperationException(s"Unsupported evaluator $name!")
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/EvaluatorType.scala]: Minor: this feels more like an IllegalArgumentException to me.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala [hunk 2]
```diff
@@ -28,6 +28,8 @@ protected[ml] class PoissonLossEvaluator(
     labelAndOffsetAndWeights: RDD[(Long, (Double, Double, Double))],
     defaultScore: Double = 0.0) extends Evaluator {
 
+  protected val evaluatorType = LogisticLoss
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala]: Should be ```PoissonLoss```.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/SmoothedHingeLossEvaluator.scala [hunk 3]
```diff
@@ -26,7 +26,9 @@ import com.linkedin.photon.ml.function.SmoothedHingeLossFunction
  */
 protected[ml] class SmoothedHingeLossEvaluator(
   labelAndOffsetAndWeights: RDD[(Long, (Double, Double, Double))],
-  defaultScore: Double = 0.0) extends Evaluator {
+  defaultScore: Double = 0.0) extends Evaluator{
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/SmoothedHingeLossEvaluator.scala]: Minor: missing space between ```Evaluator``` and ```{```.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/RMSEEvaluator.scala [hunk 4]
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.rdd.RDD
   */
 protected[ml] class RMSEEvaluator(
     labelAndOffsetAndWeights: RDD[(Long, (Double, Double, Double))],
-    defaultScore: Double = 0.0) extends Evaluator {
+    defaultScore: Double = 0.0) extends Evaluator(RMSE) {
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/RMSEEvaluator.scala]: Isn't this the ```Evaluator``` trait? How are we passing constructor arguments here?

I think it should be like the others:

```scala
    defaultScore: Double = 0.0) extends Evaluator {

    protected val evaluatorType = RMSE
```

File: photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/ScoreProcessingUtils.scala [hunk 5]
```diff
@@ -45,8 +45,15 @@ object ScoreProcessingUtils {
       val uid = Option(scoreAvro.getUid).map(_.toString)
       val label = Option(scoreAvro.getLabel()).map(_.toDouble)
       val ids = scoreAvro.getMetadataMap().asScala.map { case (k, v) => (k.toString, v.toString) }.toMap
+      val idsWithUid =
+        if (uid.isDefined) {
+          ids + (DefaultFieldNames.UID -> uid.get)
+        } else {
+          ids
+        }
+      uid.foreach{ id => }
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/ScoreProcessingUtils.scala]: Missing space between ```foreach``` and ```{```.

Also, are we forcing evaluation here? Should include a comment that clarifies the purpose.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/ScoreProcessingUtils.scala [hunk 6]
```diff
@@ -45,8 +45,15 @@ object ScoreProcessingUtils {
       val uid = Option(scoreAvro.getUid).map(_.toString)
       val label = Option(scoreAvro.getLabel()).map(_.toDouble)
       val ids = scoreAvro.getMetadataMap().asScala.map { case (k, v) => (k.toString, v.toString) }.toMap
+      val idsWithUid =
+        if (uid.isDefined) {
+          ids + (DefaultFieldNames.UID -> uid.get)
+        } else {
+          ids
+        }
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/ScoreProcessingUtils.scala]: I believe this way is preferred:

```scala
val idsWithUid = uid match {
  case Some(id) => ids + (DefaultFieldNames.UID -> id)
  case _ => ids
}
```

File: photon-ml/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataSet.scala [hunk 7]
```diff
@@ -36,7 +36,7 @@ import com.linkedin.photon.ml.constants.StorageLevel
  * @param featureShardId the feature shard id
  * @author xazhang
  */
-protected[ml] class RandomEffectDataSet(
+protected[ml] class   RandomEffectDataSet(
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/data/RandomEffectDataSet.scala]: Minor: extra space.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/Coordinate.scala [hunk 8]
```diff
@@ -40,7 +40,9 @@ protected[ml] abstract class Coordinate[D <: DataSet[D], C <: Coordinate[D, C]](
     */
   protected[algorithm] def initializeModel(seed: Long): DatumScoringModel
 
-  protected[algorithm] def updateModel(model: DatumScoringModel, score: KeyValueScore): (DatumScoringModel, OptimizationTracker) = {
+  protected[algorithm] def updateModel(model: DatumScoringModel, score: KeyValueScore)
+  : (DatumScoringModel, OptimizationTracker) = {
+
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/Coordinate.scala]: Should be this way:

```scala
protected[algorithm] def updateModel(
    model: DatumScoringModel,
    score: KeyValueScore): (DatumScoringModel, OptimizationTracker) = {
```

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala [hunk 9]
```diff
@@ -28,7 +28,7 @@ protected[ml] class PoissonLossEvaluator(
     labelAndOffsetAndWeights: RDD[(Long, (Double, Double, Double))],
     defaultScore: Double = 0.0) extends Evaluator {
 
-  protected val evaluatorType = LogisticLoss
+  protected val evaluatorType = PoissonLoss
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/PoissonLossEvaluator.scala]: Ah, you already fixed this. Ignore my previous comment. :-)

File: photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/RMSEEvaluator.scala [hunk 10]
```diff
@@ -26,9 +26,11 @@ import org.apache.spark.rdd.RDD
   */
 protected[ml] class RMSEEvaluator(
     labelAndOffsetAndWeights: RDD[(Long, (Double, Double, Double))],
-    defaultScore: Double = 0.0) extends Evaluator(RMSE) {
+    defaultScore: Double = 0.0) extends Evaluator {
 
-  val squaredLossEvaluator = new SquaredLossEvaluator(labelAndOffsetAndWeights, defaultScore)
+  protected val evaluatorType = RMSE
```
Comment (joshvfleming) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/evaluation/RMSEEvaluator.scala]: Also ignore my previous comment about this.

**Comment by joshvfleming** [comment]
LGTM. Nice factoring, and the ```randomEffectId -> idType``` really improves readability, IMO.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala [hunk 11]
```diff
@@ -23,20 +23,21 @@ import com.linkedin.photon.ml.{BroadcastLike, RDDLike}
...
+ *
+ * @param coordinates The individual optimization problem coordinates. The coordinates is a [[Seq]] consists of
+ *                    (coordinateName, [[Coordinate]] object) pairs.
+ * @param trainingLossFunctionEvaluator Training loss function evaluator
+ * @param validatingDataAndEvaluatorsOption Optional validation data evaluator. The validating data is a [[RDD]]
+ *                                         consists of (global Id, [[GameDatum]] object pairs), there the global Id
+ *                                         is a unique identifier for each [[GameDatum]] object. The evaluators are
+ *                                         a [[Seq]] of evaluators
+ * @param logger A logger instance
+ */
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala]: Some typos/inconsistencies here:

"... The coordinates are a [[Seq]] of (coordinateName, [[Coordinate]] object) pairs."
"... The validating data is a [[RDD]] of (globalId, [[GameDatum]] object) pairs, where globalId is a unique identifier for each [[GameDatum]] object."

File: photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/DataProcessingUtils.scala [hunk 12]
```diff
@@ -36,94 +36,71 @@ object DataProcessingUtils {
...
+      (shardId, featureMapLoader.indexMapForDriver().featureDimension)
     }
   }
 
-  //TODO: Change the scope to protected[avro] after Avro related classes/functIons are decoupled from the rest of code
+  /**
+   * Parse a [[RDD]] of type [[GameDatum]] from a [[RDD]] of type [[GenericRecord]]
+   * @param records a [[RDD]] of type [[GenericRecord]]
+   * @param featureShardIdToFeatureSectionKeysMap a map from feature shard id (defined by the user) to feature
+   *                                              section keys (define in input data's Avro schema)
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/DataProcessingUtils.scala]: "... (defined in the input data's Avro schema)"

File: photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/DataProcessingUtils.scala [hunk 13]
```diff
@@ -36,94 +36,71 @@ object DataProcessingUtils {
...
+   * @param featureShardIdToFeatureMapLoader a map from feature shard id (defined by the user) to the feature map loader
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/DataProcessingUtils.scala]: This doc doesn't offer any more information than the parameter name. What is a FeatureMapLoader?

File: photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/DataProcessingUtils.scala [hunk 14]
```diff
@@ -132,19 +109,34 @@ object DataProcessingUtils {
...
-      (randomEffectId, finalIdValue.toString)
+      (idType, finalIdValue.toString)
     }.toMap
   }
 
+  /**
+   * Parse a [[GameDatum]] from a [[GenericRecord]]
+   * @param record an instance of [[GenericRecord]]
+   * @param featureShardIdToFeatureSectionKeysMap a map from feature shard id (defined by the user) to feature
+   *                                              section keys (define in input data's Avro schema)
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/DataProcessingUtils.scala]: Same as above

File: photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/scoring/Driver.scala [hunk 15]
```diff
@@ -76,43 +74,48 @@ class Driver(val params: Params, val sparkContext: SparkContext, val logger: Pho
...
   }
 
   /**
-    * Score the game data set with the game model
-    *
-    * @param featureShardIdToFeatureMapMap A map of shard id to feature map
-    * @param gameDataSet The game data set
-    * @return The scores
-    */
+   * Score the game data set with the game model
```
Comment (ashelkovnykov) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/cli/game/scoring/Driver.scala]: Missing newline between description and params

**Comment by ashelkovnykov** [comment]
Forgot to do this yesterday - my bad. LGTM

PS. I need to hurry with the verifiers/photon-client proposal or else the Params will just keep growing...

File: photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/ScoreProcessingUtils.scala (hunk 5 above)
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/avro/data/ScoreProcessingUtils.scala]: Looks like the diff of this file got somehow messed up. Will fix it.

File: photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala (hunk 11 above)
Comment (XianXing) [inline photon-ml/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateDescent.scala]: Good catch, thanks!

**Pull request merged**